rtCommon/bidsArchive.py

```diff
@@ -654,3 +655,78 @@ def getIncremental(self, imageIndex: int = 0, **entities) \
         except MissingMetadataError as e:
             raise MissingMetadataError("Archive lacks required metadata for "
                                        "BIDS Incremental creation: " + str(e))
+
+    @failIfEmpty
+    def getBidsRun(self, **entities) -> BidsRun:
+        """
+        Get a BIDS Run from the archive.
+
+        Args:
+            entities: Entities defining a run in the archive.
+
+        Returns:
+            A BidsRun containing all the BidsIncrementals in the specified run.
+
+        Raises:
+            NoMatchError: If the entities don't match any runs in the archive.
+            QueryError: If the entities match more than one run in the archive.
+
+        Examples:
+            >>> archive = BidsArchive('/tmp/dataset')
+            >>> run = archive.getBidsRun(subject='01', session='02',
+                                         task='testTask', run=1)
+            >>> print(run.numIncrementals())
+            53
+        """
+        images = self.getImages(**entities)
+        if len(images) == 0:
+            raise NoMatchError(f"Found no runs matching entities {entities}")
+        if len(images) > 1:
+            entities = [img.get_entities() for img in images]
+            raise QueryError("Provided entities were not unique to one run; "
+                             "try specifying more entities "
+                             f" (got runs with these entities: {entities}")
+        else:
+            bidsImage = images[0]
+            niftiImage = bidsImage.get_image()
+            niftiData = getNiftiData(niftiImage)
+
+            metadata = self.getSidecarMetadata(bidsImage)
+            metadata.pop('extension')  # only used in PyBids
+
+            run = BidsRun(**bidsImage.get_entities())
+            numImages = niftiData.shape[3]
+            for imageIdx in range(numImages):
+                newData = niftiData[..., imageIdx]
+                newImage = nib.Nifti1Image(newData,
+                                           niftiImage.affine,
+                                           niftiImage.header)
+
+                newIncremental = BidsIncremental(newImage, metadata)
+                run.appendIncremental(newIncremental, validateAppend=False)
```

Comment:
Can BidsRun __init__() function have an optional parameter which is the niftiData to load from? To avoid making BidsIncrementals for each volume. Or maybe a loadFromBidsImage() function?

Reply:
Since all of the BIDS Incremental data is needed (entities, sidecar metadata, NIfTI data, NIfTI header), I'd prefer to leave the interface in terms of the BIDS-Incremental to avoid having a BidsRun that could be in an inconsistent state (e.g., has niftiData but doesn't have sidecar metadata or entities).

However, because BidsRun's  `appendIncremental`  splits up the data in a BidsIncremental into its component arrays, this processing code can be removed and that routine can handle it. (See BidsRun:155-161)

Reply:
Ok yes, makes sense.